File: message.py
import unittest
from enum import Enum


class Type(Enum):
    REQ = int('011', 2)
    REQ_M = int('010', 2)
    APR = int('100', 2)
    NACK = int('101', 2)
    DATA = int('000', 2)
    KEEP_A = int('110', 2)


class Mask(Enum):
    #            checksum, hash, window, filename
    REQ = [0b111, 0x1FFFFF, (1 << 256) - 1, 0xFF, '_']
    #            checksum, window
    REQ_M = [0b111, 0x1FFFFF, 0xFF, '_']
    #            checksum, seq_number
    APR = [0b111, 0x1FFFFF, (1 << 32) - 1]
    #            checksum, seq_number
    NACK = [0b111, 0x1FFFFF, (1 << 32) - 1]
    #            checksum, seq_number, data
    DATA = [0b111, 0x1FFFFF, (1 << 32) - 1, '_']
    #
    KEEP_A = [0b111, 0x1FFFFF]


def unpack_message(message_bytes):
    type = Type(message_bytes[0] >> 5)

    message_mask = Mask[type.name].value

    message_bits = ''.join(format(byte, '08b') for byte in message_bytes)

    fields = []
    for index, mask_field in enumerate(message_mask):
        if mask_field == '_':
            fields.append(int(message_bits, 2))
            break

        fields.append(int(message_bits[:mask_field.bit_length()], 2))
        message_bits = message_bits[mask_field.bit_length():]

    return fields


def pack_message(type, *fields):
    message_type = Type(type)

    message_mask = Mask[message_type.name].value

    fields = (type,) + fields

    message_bits = ''
    for index, mask_field in enumerate(message_mask):

        if mask_field == '_':
            message_bits += bin(fields[index])[2:].zfill(((fields[index].bit_length() + 7) // 8) * 8)
            break

        message_bits += bin(fields[index] & mask_field)[2:].zfill(mask_field.bit_length())

    return bytes([int(message_bits[i:i + 8], 2) for i in range(0, len(message_bits), 8)])


def compute_checksum(bits_length, *data):
    checksum = sum(data)

    while checksum.bit_length() > bits_length:
        # Split the checksum into two halves
        mask = (1 << (checksum.bit_length() // 2)) - 1
        low_bits = checksum & mask
        high_bits = checksum >> (checksum.bit_length() // 2)

        checksum = low_bits + high_bits

    return checksum


def create_message(type, *fields):
    checksum = compute_checksum(21, *((type,) + fields))
    fields = (checksum,) + fields
    return pack_message(type, *fields)


def open_message(message):
    fields = unpack_message(message)
    checksum = fields.pop(1)
    if checksum != compute_checksum(21, *fields):
        raise ValueError('Invalid checksum')
    return fields

def corrupt_message(message):
     return message[:1] + bytes([ord('a')]) + message[2:]

class Message(unittest.TestCase):
    def test__compute_checksum(self):
        fields = [
            Type.REQ.value,
            0x3398d424e5d1a1f2657dc06680dee743aced01dd177b69c32a09c70f1f362bb9,
            1, int('<3 U )'.encode().hex(), 16)
        ]
        self.assertEqual(compute_checksum(8, *fields), 31)
        self.assertEqual(compute_checksum(21, *fields), 173013)

    def test__pack_message_01(self):
        fields = [
            Type.REQ.value, 0xffff,
            0x3398d424e5d1a1f2657dc06680dee743aced01dd177b69c32a09c70f1f362bb9,
            1, int('<3 U )'.encode().hex(), 16)
        ]
        message = pack_message(*fields)
        _fields = unpack_message(message)
        print(_fields[-1].to_bytes((_fields[-1].bit_length() + 7) // 8, 'big').decode())
        self.assertEqual(fields, _fields)

    def test__create_message_01(self):
        fields = [
            Type.REQ.value,  # without checksum
            0x3398d424e5d1a1f2657dc06680dee743aced01dd177b69c32a09c70f1f362bb9,
            1, int('<3 U )'.encode().hex(), 16)
        ]
        message = create_message(*fields)
        _fields = unpack_message(message)
        self.assertEqual(173013, _fields[1])

    def test__open_message(self):
        fields = [
            Type.REQ.value,  # without checksum
            0x3398d424e5d1a1f2657dc06680dee743aced01dd177b69c32a09c70f1f362bb9,
            1, int('<3 U )'.encode().hex(), 16)
        ]
        message = create_message(*fields)
        _fields = open_message(message)
        self.assertEqual(fields, _fields)

        message = message[:1] + bytes([ord('a')]) + message[2:]
        with self.assertRaises(ValueError):
            _fields = open_message(message)

    if __name__ == '__main__':
        unittest.main()

File: message.js
/**
 * Message Module - SNSS Protocol
 * Handles message packing, unpacking, and checksum validation
 */

/**
 * Message Types - 3 bits
 */
const Type = {
    REQ: 0b011,      // File request
    REQ_M: 0b010,    // Message request
    APR: 0b100,      // Acknowledgment
    NACK: 0b101,     // Negative acknowledgment
    DATA: 0b000,     // Data chunk
    KEEP_A: 0b110    // Keep-alive
};

/**
 * Field masks for each message type
 * '_' indicates variable-length field (rest of message)
 */
const Mask = {
    // checksum, hash, window, filename
    REQ: [0b111, 0x3FFFF, 0xFF, 0x7FF, '_'],
    // checksum, window, payload
    REQ_M: [0b111, 0x3FFFF, 0xFF, 0x7FF],
    // checksum, seq_number
    APR: [0b111, 0x1FFFFF, 0xFFFFFFFF],
    // checksum, seq_number
    NACK: [0b111, 0x1FFFFF, 0xFFFFFFFF],
    // checksum, seq_number, data
    DATA: [0b111, 0x1FFFFF, 0xFFFFFFFF, '_'],
    // checksum
    KEEP_A: [0b111, 0x1FFFFF]
};

/**
 * Get type name from value
 */
function getTypeName(typeValue) {
    for (const [name, value] of Object.entries(Type)) {
        if (value === typeValue) return name;
    }
    throw new Error(`Unknown type value: ${typeValue}`);
}

/**
 * Get bit length of a number
 */
function bitLength(num) {
    if (num === 0) return 1;
    return Math.floor(Math.log2(num)) + 1;
}

/**
 * Convert number to binary string with zero padding
 */
function toBinaryString(num, length) {
    return num.toString(2).padStart(length, '0');
}

/**
 * Unpack message from bytes
 */
function unpackMessage(messageBytes) {
    // Extract type from first 3 bits
    const typeValue = messageBytes[0] >> 5;
    const typeName = getTypeName(typeValue);
    const messageMask = Mask[typeName];
    
    // Convert bytes to binary string
    let messageBits = '';
    for (let i = 0; i < messageBytes.length; i++) {
        messageBits += toBinaryString(messageBytes[i], 8);
    }
    
    const fields = [];
    
    for (let index = 0; index < messageMask.length; index++) {
        const maskField = messageMask[index];
        
        if (maskField === '_') {
            // Variable length field - rest of message
            if (messageBits.length > 0) {
                fields.push(BigInt('0b' + messageBits));
            } else {
                fields.push(BigInt(0));
            }
            break;
        }
        
        const maskBitLength = bitLength(maskField);
        const fieldBits = messageBits.substring(0, maskBitLength);
        fields.push(parseInt(fieldBits, 2));
        messageBits = messageBits.substring(maskBitLength);
    }
    
    return fields;
}

/**
 * Pack message fields into bytes
 */
function packMessage(type, ...fields) {
    const typeName = getTypeName(type);
    const messageMask = Mask[typeName];
    
    const allFields = [type, ...fields];
    let messageBits = '';
    
    for (let index = 0; index < messageMask.length; index++) {
        const maskField = messageMask[index];
        
        if (maskField === '_') {
            // Variable length field
            const fieldValue = allFields[index];
            const fieldBigInt = typeof fieldValue === 'bigint' ? fieldValue : BigInt(fieldValue);
            
            if (fieldBigInt === BigInt(0)) {
                break;
            }
            
            const fieldBitLength = Math.ceil((fieldBigInt.toString(2).length) / 8) * 8;
            messageBits += fieldBigInt.toString(2).padStart(fieldBitLength, '0');
            break;
        }
        
        const fieldValue = allFields[index] & maskField;
        messageBits += toBinaryString(fieldValue, bitLength(maskField));
    }
    
    // Convert binary string to bytes
    const bytes = [];
    for (let i = 0; i < messageBits.length; i += 8) {
        const byte = messageBits.substring(i, i + 8);
        bytes.push(parseInt(byte, 2));
    }
    
    return Buffer.from(bytes);
}

/**
 * Compute checksum using Internet Checksum algorithm
 */
function computeChecksum(bitsLength, ...data) {
    let checksum = BigInt(0);
    
    for (const value of data) {
        checksum += typeof value === 'bigint' ? value : BigInt(value);
    }
    
    while (bitLength(Number(checksum)) > bitsLength) {
        const checksumBitLength = bitLength(Number(checksum));
        const halfLength = Math.floor(checksumBitLength / 2);
        const mask = (BigInt(1) << BigInt(halfLength)) - BigInt(1);
        
        const lowBits = checksum & mask;
        const highBits = checksum >> BigInt(halfLength);
        
        checksum = lowBits + highBits;
    }
    
    return Number(checksum);
}

/**
 * Create message with checksum
 */
function createMessage(type, ...fields) {
    const checksum = computeChecksum(21, type, ...fields);
    const fieldsWithChecksum = [checksum, ...fields];
    return packMessage(type, ...fieldsWithChecksum);
}

/**
 * Open and validate message
 */
function openMessage(message) {
    const fields = unpackMessage(message);
    const checksum = fields[1];
    
    // Remove checksum for validation
    const fieldsWithoutChecksum = [fields[0], ...fields.slice(2)];
    const calculatedChecksum = computeChecksum(21, ...fieldsWithoutChecksum);
    
    if (checksum !== calculatedChecksum) {
        throw new Error('Invalid checksum');
    }
    
    // Return fields without checksum
    return fieldsWithoutChecksum;
}

/**
 * Corrupt message for testing
 */
function corruptMessage(message) {
    const corrupted = Buffer.from(message);
    if (corrupted.length > 1) {
        corrupted[1] = 'a'.charCodeAt(0);
    }
    return corrupted;
}

module.exports = {
    Type,
    Mask,
    unpackMessage,
    packMessage,
    computeChecksum,
    createMessage,
    openMessage,
    corruptMessage,
    getTypeName
};
